scripts/notify-new-listings.js

```diff
@@ -0,0 +1,126 @@
+  // for local testing and development.
+  baseUrl: process.env.FLEX_INTEGRATION_BASE_URL || 'https://flex-integ-api.sharetribe.com',
+});
+
+// Start polloing from current time on, when there's no stored state
+const startTime = new Date();
+
+// Polling interval (in ms) when all events have been fetched.
+// Keeping this at 10s or more is a good idea.
+const pollIdleWait = 10000;
+// Polling interval (in ms) when a full page of events is received and there may be more
+const pollWait = 250;
+
+// File to keep state across restarts. Stores the last seen event sequence ID,
+// which allows continuing polling from the correct place
+const stateFile = "./notify-new-listings.state";
+
+const queryEvents = (args) => {
+  var filter = {eventTypes: "listing/created,listing/updated"};
+  return integrationSdk.events.query(
+    {...args, ...filter}
+  );
+};
+
+const saveLastEventSequenceId = (sequenceId) => {
+  try {
+    fs.writeFileSync(stateFile, sequenceId);
+  } catch (err) {
+    throw err;
+  }
+};
+
+const loadLastEventSequenceId = () => {
+  try {
+    const data = fs.readFileSync(stateFile);
+    return parseInt(data, 10);
+  } catch (err) {
+    return undefined;
+  }
+};
+
+const analyzeEvent = (event) => {
+  if (event.attributes.resourceType == "listing") {
+    const listing = event.attributes.resource;
+    const listingId = event.attributes.resourceId.uuid;
+    const authorId = listing.relationships.author.data.id.uuid;
+    const listingState = listing.attributes.state;
+    const previousValues = event.attributes.previousValues;
+    if (event.attributes.eventType == "listing/created") {
+      switch(listingState) {
+      case "pendingApproval":
+        console.log(`A new listing is pending approval: listing ID ${listingId}, author ID: ${authorId}`);
+        break;
+      case "published":
+        console.log(`A new listing has been published: listing ID ${listingId}, author ID: ${authorId}`);
+        break;
+      };
+    } else if (event.attributes.eventType == "listing/updated") {
+      // listing is approved when the new state is published and the previous
+      // state was pendingApproval
+      if (listingState == "published"
+          && previousValues.attributes
+          && previousValues.attributes.state == "pendingApproval") {
+        console.log(`Listing has been approved by operator: listing ID ${listingId}, author ID: ${authorId}`);
+      } else if (listingState == "published"
+                 && previousValues.attributes
+                 && previousValues.attributes.state == "draft") {
+        console.log(`A new listing has been published: listing ID ${listingId}, author ID: ${authorId}`);
+      } else if (listingState == "pendingApproval"
+                 && previousValues.attributes
+                 && previousValues.attributes.state == "draft") {
+        console.log(`A new listing is pending approval: listing ID ${listingId}, author ID: ${authorId}`);
+      }
+    }
```

Comment:
This is a bit hard to read. How about:

```js
const { 
  resourceId, 
  resource: listing,
  state: listingState,
  previousValues,
  eventType,
} = event.attributes;
const listingId = resourceId.uuid;
const authorId = listing.relationships.author.data.id.uuid;
const listingDetails = `listing ID ${listingId}, author ID: ${authorId}`;
const { state: previousState } = previousValues.attributes || {};

const isPublished = listingState == "published";
const isPendingApproval = listingState == "pendingApproval";
const wasDraft = previousState == "published";
const wasPendingApproval = previousState == "pendingApproval";

switch(eventType) {
  case "listing/created": 
    if (isPendingApproval) {
      console.log(`A new listing is pending approval: ${listingDetails}`);
    } else if (isPublished) {
      console.log(`A new listing has been published: ${listingDetails}`);
    }
    break;
  case "listing/updated":
    if (isPublished && wasPendingApproval) {
      console.log(`Listing has been approved by operator: ${listingDetails}`);
    } else if (isPublished && wasDraft) {
      console.log(`A new listing has been published: ${listingDetails}`);
    } else if (isPendingApproval && wasDraft) {
      console.log(`A new listing is pending approval: ${listingDetails}`);
    }
    break;
  default:
    console.log(`Unhandled eventType: ${eventType}`);
}
```

Disclaimer: untested code.

Reply:
Uh, my code contained at least one bug: 
`const wasDraft = previousState === "published";`
`const wasDraft = previousState === "draft";`